T20146 copy usb apps

treitter opened the pull request:
This implements copying Flatpaks to USB through a new button on the details page.

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -2334,6 +2559,22 @@ gs_details_page_setup (GsPage *page,
+	for (l = mounts; l; l = l->next) {
+		gs_details_page_mount_added_cb (self->volume_monitor, l->data,
+						self);
+	}
```
Nitpick: No need for curly brackets.

joaquimrocha replied:
I also think this is too much responsibility for the details view. Usually views in g-s shouldn't have so much feature-related logic. Instead, this should be in a plugin, which would affect the apps' properties accordingly (like we do for the shortcuts). The EOS plugin also already tracks mountpoints, so this logic should be there, and maybe we just add a `COPIABLE` prop to the app (or do it using the app's metadata if easier), so the details views (and others maybe) will be notified if an app is copiable or not.

joaquimrocha commented on src/gs-page.c:
```diff
@@ -246,6 +250,36 @@ gs_page_app_removed_cb (GObject *source,
+	if (g_error_matches (error,
+			     GS_PLUGIN_ERROR,
+			     GS_PLUGIN_ERROR_CANCELLED)) {
+		g_debug ("%s", error->message);
```
Maybe this message needs a bit more context to be useful.

treitter replied:
I prefixed with "app copy cancelled: "

joaquimrocha commented on src/gs-page.c:
```diff
@@ -246,6 +250,36 @@ gs_page_app_removed_cb (GObject *source,
+
+	if (GS_PAGE_GET_CLASS (page)->app_copied != NULL) {
+		GS_PAGE_GET_CLASS (page)->app_copied (page, helper->app);
+	}
```
Nitpick: No need for curly brackets.

treitter replied:
Fixed locally

joaquimrocha commented on src/gs-page.c:
```diff
@@ -481,6 +515,45 @@ gs_page_install_app (GsPage *page,
+	GsPagePrivate *priv = gs_page_get_instance_private (page);
+	GsPageHelper *helper;
+
+	helper = g_slice_new0 (GsPageHelper);
```
Nitpick: You can initialize it together with the declaration. And use `g_new0` is probably better too.

treitter replied:
I've merged the lines locally.

But the free function uses `g_slice_free()` and all the other instances of `GsPageHelper` allocate it with `g_slice_new0()` so we'd need to change those as well and I don't see a critical need to do that.

Why do you prefer `g_new0()`? Because it's less of a risk when debugging and the performance gains of `g_slice` allocation are trivial here?

joaquimrocha replied:
I read somewhere that `g_slice` was no longer a faster solution. But I hadn't noticed the free func, so let's keep it then.

joaquimrocha commented on src/gs-page.c:
```diff
@@ -481,6 +515,45 @@ gs_page_install_app (GsPage *page,
+	helper->cancellable = g_object_ref (cancellable);
+	helper->interaction = interaction;
+
+	if (gs_app_get_state (app) == AS_APP_STATE_INSTALLED) {
```
Use `gs_app_is_installed`. Otherwise, if the app is e.g. updatable, then it won't be able to copy it to the USB.

treitter replied:
Good catch. Fixed locally

joaquimrocha commented on src/gs-page.c:
```diff
@@ -481,6 +515,45 @@ gs_page_install_app (GsPage *page,
+						    gs_page_app_copied_cb,
+						    helper);
+	} else {
+		// NOTE: it should be impossible to reach this state
```
I think it should be unlikely this is reached, but the states here are not atomic, so things like this are usually reachable in g-s :sweat_smile:

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+{
+	if (copyable) {
+		*copyable = (gs_flatpak_get_collection_id (self, app) != NULL);
+	}
```
Nitpick: No need for curly brackets.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+gs_flatpak_copy_process_helper_free (GsFlatpakCopyProcessHelper *helper)
+{
+	g_object_unref (helper->app);
+	g_object_unref (helper->cancellable);
```
Nitpick: Maybe it's better/safer to clear it these objects instead?

treitter replied:
Sure. Fixed locally.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+copy_process_watch_cb (GPid pid, gint status, gpointer user_data)
+{
+	GsFlatpakCopyProcessHelper *helper =
+		(GsFlatpakCopyProcessHelper*) user_data;
```
No need for the cast.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+		(GsFlatpakCopyProcessHelper*) user_data;
+	g_autoptr(GError) error = NULL;
+
+	if (!g_cancellable_is_cancelled (helper->cancellable) && status) {
```
Nitpick: I'd explicitly check if the status != 0.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+			     "failed to copy flatpak app to removable "
+			     "media: `flatpak create-usb` failed with "
+			     "status %d", status);
+	}
```
Nitpick: No need for curly brackets.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	if (error != NULL) {
+		g_warning ("failed to refresh flatpak metadata after copying "
+			"app to USB: %s", error->message);
+	}
```
Nitpick: No need for curly brackets.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+			    NULL, &error);
+	if (error != NULL) {
+		g_warning ("failed to refresh flatpak metadata after copying "
+			"app to USB: %s", error->message);
```
Alignment looks odd(er than usual), but maybe it's still Github's view.

treitter replied:
it was one tab in from the start of `g_warning` so I added the missing spaces locally.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	// update metadata to ensure the USB category includes this new app and
+	// we can immediately download an app, copy to USB, delete from
+	// computer, and install from USB without leaving the app page
+	gs_flatpak_refresh (helper->self, 0, GS_PLUGIN_REFRESH_FLAGS_METADATA,
```
Are you sure this is the best way to do this? I think the refresh lists the apps in the repo which is quite slow, so maybe we should just modify the app as needed?

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+
+	// update metadata to ensure the USB category includes this new app and
+	// we can immediately download an app, copy to USB, delete from
+	// computer, and install from USB without leaving the app page
```
Use C style comments.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+{
+	GPid pid = GPOINTER_TO_INT (user_data);
+
+	// terminate the process which is copying the app
```
Use C style comments.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+		     GError **error)
+{
+	// this is used in an async function but we block here until that
+	// returns so we won't auto-free while other threads depend on this
```
Use C style comments.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	app_id = gs_flatpak_app_get_ref_name (app);
+	app_arch = gs_flatpak_app_get_ref_arch (app);
+	app_branch = gs_flatpak_app_get_ref_branch (app);
+	app_ref = g_strdup_printf ("%s/%s/%s", app_id,
```
Nitpick: Maybe set all these together with their declaration?

treitter replied:
I thought it was convention to have no assignments (except for `NULL`) at the beginning of a block but maybe I was confusing `-Wdeclaration-after-statement` (which doesn't mind assignments mixed into the declaration section after all).

Fixed locally.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+						     app_arch, app_branch);
+	app_collection_id = gs_flatpak_get_collection_id (self, app);
+	g_debug ("Copying app %s with collection ID: %s", app_ref,
+		 (app_collection_id ? app_collection_id : "(null)"));
```
I think that in `g_debug` a NULL string arg will already show up as `(null)`.

treitter replied:
I think you're right.

Fixed.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	g_debug ("Copying app %s with collection ID: %s", app_ref,
+		 (app_collection_id ? app_collection_id : "(null)"));
+
+	argv = g_malloc (sizeof (gchar*) * 5);
```
You should declare a constant `GStrv` here. See https://github.com/endlessm/gnome-software/blob/master/src/gs-category-page.c#L411 .

treitter replied:
Good idea.

Fixed locally.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+							     cancelled_id);
+		g_child_watch_add (child_pid, copy_process_watch_cb, helper);
+	} else if (error && *error) {
+		return FALSE;
```
I think the return value should be dependent on the passed error arg.

treitter replied:
Do you mean independent? Because that makes sense.

Fixed locally.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	// user cancels the copy. Without this, it is impossible to cancel the
+	// copy because we reach the end of this function, its parent GTask
+	// returns and we disconnect the handler that would kill the copy
+	// process.
```
Use C style comments.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	// process.
+	while (!helper->finished) {
+		g_main_context_iteration (NULL, FALSE);
+	}
```
Nitpick: No need for curly brackets.

treitter replied:
Fixed locally

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,182 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+	// returns and we disconnect the handler that would kill the copy
+	// process.
+	while (!helper->finished) {
+		g_main_context_iteration (NULL, FALSE);
```
This will be executed from a different thread. Is this fine?

treitter replied:
I think that's fine as long as we block here until `copy_process_watch_cb()` executes and breaks the conditional which is what I intended.

Do you see any potential races?

joaquimrocha replied:
I don't know exactly. I'd have to probe it, just wanted to make sure this is checked for races.

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -446,31 +446,28 @@ gs_details_page_refresh_progress (GsDetailsPage *self)
 	}
 	if (app_has_pending_action (self->app)) {
 		GsPluginAction action = gs_app_get_pending_action (self->app);
+
```
Why the extra newline in this commit?

treitter replied:
No reason. Fixed locally.

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -446,31 +446,28 @@ gs_details_page_refresh_progress (GsDetailsPage *self)
 		break;
 	}
-	if (app_has_pending_action (self->app)) {
+	if (plugin_or_app_has_pending_action (self, self->app)) {
```
Commit titles in g-s (usually) only start by the module when it's a change in the plugins. Also seems like this commit should rather be squashed into the others you did before?

treitter replied:
Fixed locally.

This conditional only needs to change when we support copying so I don't think it's necessarily better to move the support function to the previous commit given it isn't used yet by then (and would trigger an "unused function" warning when building at that point).

joaquimrocha commented:
Reviewed and wrote a few comments about the code and nitpicks. I think the approach here is putting too much feature-related code into the details view, instead of using our plugin.

treitter commented on lib/gs-plugin-job.c:
```diff
@@ -67,6 +68,7 @@ enum {
 	PROP_MAX_RESULTS,
 	PROP_PRICE,
 	PROP_TIMEOUT,
+	PROP_DEST,
```
I've fixed the name locally.

I want the flexibility to change the destination on a per-job basis, I think, since the path can change as USB disks are added and removed and, in the future, I'll want to add the ability for the user to select the destination. Right now, it just picks the most-recently-mounted removable disk which is kind of crude.

Given those needs, would it still make sense to do this in the plugin loader?

treitter commented on lib/gs-plugin-job.c:
```diff
@@ -103,6 +105,10 @@ gs_plugin_job_to_string (GsPluginJob *self)
+	if (self->dest != NULL) {
+		g_string_append_printf (str, " with destination=%s",
+					self->dest);
+	}
```
I changed this to "copy destination" locally.

treitter commented on lib/gs-plugin-loader.c:
```diff
@@ -1674,6 +1704,100 @@ gs_plugin_loader_job_get_categories_finish (GsPluginLoader *plugin_loader,
+
+	/* save helper */
+	helper = gs_plugin_loader_helper_new (plugin_loader, plugin_job);
+	helper->catlist = NULL;
```
I wanted to be sure we didn't try freeing uninitialized memory later but forgot I'm doing that in `gs_plugin_loader_helper_new()`, so I've removed this locally.

treitter commented on lib/gs-plugin-loader.c:
```diff
@@ -3098,6 +3222,7 @@ gs_plugin_loader_generic_update (GsPluginLoader *plugin_loader,
 			gs_plugin_loader_action_start (plugin_loader, plugin, FALSE);
 			ret = plugin_app_func (plugin, app, app_cancellable, &error_local);
 			gs_plugin_loader_action_stop (plugin_loader, plugin);
+
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -238,6 +243,105 @@ gs_details_page_update_shortcut_button (GsDetailsPage *self)
+
+	if (!copyable) {
+		gtk_widget_set_sensitive (self->button_copy, FALSE);
+		gtk_widget_set_visible (self->button_copy, FALSE);
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -238,6 +243,105 @@ gs_details_page_update_shortcut_button (GsDetailsPage *self)
+
+	if (self->plugin_status == GS_PLUGIN_STATUS_COPYING) {
+		return;
+	}
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -238,6 +243,105 @@ gs_details_page_update_shortcut_button (GsDetailsPage *self)
+			   as_app_state_to_string (state));
+		g_assert_not_reached ();
+	}
+}
```
Fixed locally with a simple call to `gs_app_is_installed()`.

treitter commented on src/gs-details-page.c:
```diff
@@ -253,6 +357,13 @@ app_has_pending_action (GsApp *app)
 }
 
+static gboolean
+plugin_or_app_has_pending_action (GsDetailsPage *self)
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -532,8 +651,8 @@ gs_details_page_refresh_progress (GsDetailsPage *self)
-	if (plugin_or_app_has_pending_action (self, self->app)) {
-		gtk_widget_set_visible (self->box_progress, TRUE);
+	if (plugin_or_app_has_pending_action (self)) {
+	 	gtk_widget_set_visible (self->box_progress, TRUE);
```
The perils of mixing tabs and spaces :)

Fixed locally.

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+
+		if (g_strcmp0 (uuid_needle, uuid_cur) == 0) {
+			return l->data;
+		}
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+	GList *l;
+	g_autofree gchar* uuid_needle = g_mount_get_uuid (needle);
+
+	for (l = mounts; l; l = l->next) {
```
I did that originally but saw other code that did it this way. So, I was actually trying to follow upstream's style :) I guess I just didn't look for enough examples.

Fixed locally.

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+
+	if (!g_drive_is_removable (drive)) {
+		return;
+	}
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
 	gs_shell_modal_dialog_present (self->shell, GTK_DIALOG (dialog));
 }
 
+static gconstpointer
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+	g_autoptr(GDrive) drive = g_mount_get_drive (mount);
+	g_autoptr(GFile) root = g_mount_get_root (mount);
+	g_autofree gchar *path = g_file_get_path (root);
+	g_autoptr(GMount) mount_existing = NULL;
```
I usually name variables `noun_adjective` since the noun is usually more important and it makes similar variables easier to visually group (eg, if we were re-mounting a device in a new location and had "mount_existing" and "mount_new").

But it's also not a big deal to me and it matters less when there's a single variable.

Fixed locally.

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+	}
+
+	mount_existing = (GMount*) find_mount_in_list (self->removable_mounts,
+						       mount);
```
Fixed locally

treitter commented on src/gs-details-page.c:
```diff
@@ -2079,6 +2210,71 @@ gs_details_page_write_review_cb (GtkButton *button,
+	if (mount_existing) {
+		self->removable_mounts = g_list_remove (self->removable_mounts,
+							mount_existing);
+	}
```
I get a little itchy when an unbraced body spans multiple lines in C but if that's the upstream style, I'll follow it.

joaquimrocha replied:
Yeah, let's change it if you don't mind.

treitter commented on src/gs-details-page.c:
```diff
@@ -2091,6 +2287,16 @@ gs_details_page_app_removed (GsPage *page, GsApp *app)
+	GsDetailsPage *self = GS_DETAILS_PAGE (page);
+
+	gtk_button_set_label (GTK_BUTTON (self->button_copy),
+			      _("Copied to USB"));
```
Not with 80-character lines. I guess upstream isn't very strict or does it just have a higher limit? 100 characters?

joaquimrocha replied:
I'd prefer if there was a limit, but there's actually no limit enforced, as long as it's "reasonable".

treitter replied:
OK. I joined the line.

treitter commented on src/gs-details-page.c:
```diff
@@ -2315,6 +2521,23 @@ gs_details_page_network_available_notify_cb (GsPluginLoader *plugin_loader,
+		gs_app_add_quirk (app, AS_APP_QUIRK_COMPULSORY);
+	} else {
+		/* allow removal once copying has completed */
+		gs_app_remove_quirk (app, AS_APP_QUIRK_COMPULSORY);
```
When I discussed making sure the app couldn't be deleted during a copy, Richard told me to be sure to set this.

The one concern I did have is that any "clean up" code that unsets this does so absolutely. So, it's like having two critical sections of code each guarded by a "mutex" that either can completely unlock on their own.

Maybe there's some reason why this isn't the case for quirks but it seems like an issue and there's no obvious way to solve it short of giving them ref counters.

joaquimrocha replied:
Yeah, achieving the same effect with a different solution would probably be too complex anyway as you say. Let's leave it then.

treitter commented:
This is ready for a re-review.

The latest substantial change is that I moved the mount monitoring code from `details-page` to `plugin-loader`.

I also squashed my trivial fixes into two fixup commits to hopefully make this next review a little easier. I'll squash them before the final merge.

As discussed, this defers relocating the mount management to plugins and reporting copyability as a quirk. We'll do that at the time we support users selecting the target destination themselves.

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -651,8 +626,8 @@ gs_details_page_refresh_progress (GsDetailsPage *self)
-	if (plugin_or_app_has_pending_action (self)) {
-	 	gtk_widget_set_visible (self->box_progress, TRUE);
+	if (plugin_has_pending_action (self)) {
+		gtk_widget_set_visible (self->box_progress, TRUE);
```
Nitpick: The curly braces.

joaquimrocha commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3566,18 +3553,16 @@ gs_flatpak_app_copy (GsFlatpak *self,
 							     cancelled_id);
 		g_child_watch_add (child_pid, copy_process_watch_cb, helper);
-	} else if (error && *error) {
+	} else
```
So, the thing about not having curly braces on one-liners is that if it's an if/else block where the counterpart has curly braces you should do it too... (Sorry if this is too nitpicky but I think this is how it usually goes upstream)

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -2185,65 +2179,21 @@ gs_details_page_write_review_cb (GtkButton *button,
-						 g_object_ref (mount));
-
-	gs_details_page_update_copy_button (self);
+	if (self->copy_dests)
```
Check if `!= NULL` (upstream style...).

joaquimrocha commented on src/gs-details-page.c:
```diff
@@ -286,9 +280,9 @@ gs_details_page_update_copy_button (GsDetailsPage *self)
+	copy_dest = removable_destination_dup (self);
 	if (gs_app_is_installed (self->app)) {
-		if (mount_path) {
+		if (copy_dest) {
```
Check if `!= NULL` (upstream style).

joaquimrocha commented on lib/gs-plugin-loader.c:
```diff
@@ -2459,6 +2463,96 @@ gs_plugin_loader_setup_again (GsPluginLoader *plugin_loader)
+	for (GList *l = mounts; l; l = l->next) {
+		g_autoptr(GFile) root = g_mount_get_root (l->data);
+
+		if (root)
```
Check if `!= NULL` (upstream style).

mwleeds commented on plugins/flatpak/gs-flatpak.c:
```diff
@@ -3396,6 +3411,167 @@ gs_flatpak_create_app_from_repo_dir (GsFlatpak *self,
+								       app);
+
+	gboolean spawn_retval;
+	const gchar *argv[] = {"/usr/bin/flatpak", "create-usb", copy_dest,
```
Isn't it bad to assume flatpak was installed into that location, especially if this will be upstreamed? GNOME Builder seems to not need to do that: https://gitlab.gnome.org/GNOME/gnome-builder/blob/master/src/plugins/flatpak/gbp-flatpak-runtime.c#L119

mwleeds replied:
Although in that case it's making sure `PATH` is set. So it might be better to use `which`.

mwleeds replied:
I still don't think this is the right approach (for example on NixOS binaries are installed into `/root/.nix-profile/bin/` rather than `/usr/bin/`) but there is precedent for it in the codebase (search for `DPKG_DEB_BINARY`) so I guess leave it and see what upstream says.
